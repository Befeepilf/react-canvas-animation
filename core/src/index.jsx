import React from 'react';

export default class Canvas extends React.Component {
  constructor(props) {
    super(props);

    this.sketch = props.sketch.bind(this);

    // lift custom prototype functions (added through @react-canvas-animation/draw) to this and bind this
    // such that each of these functions has access to properties that are specific to this instance of Canvas (e.g. ctx)
    for (const p in Canvas.prototype) {
      if (
        typeof Canvas.prototype[p] === 'function' &&
        Canvas.prototype.customFunctions &&
        Canvas.prototype.customFunctions.indexOf(p) > -1
      ) {
        this[p] = Canvas.prototype[p].bind(this);
      }
    }

    this.state = {
      width: props.width || 0,
      height: props.height || 0,
      customSize: props.width && props.height,
      ctx: null
    };

    // make access to width & height more convenient
    this.width = this.state.width;
    this.height = this.state.height;

    this.onRefChange = this.onRefChange.bind(this);
    this.onResize = this.onResize.bind(this);
    this.draw = this.draw.bind(this);
  }

  // updates ctx if the reference to the canvas element changes
  onRefChange(node) {
    if(node) {
      this.ctx = node.getContext('2d');
      this.setState({ctx: this.ctx});
    }
  }

  onResize() {
    if(!this.state.customSize) {
      this.setState({width: window.innerWidth, height: window.innerHeight});
      this.width = window.innerWidth;
      this.height = window.innerHeight;
    }
  }

  // this function draws one frame by clearing the canvas and calling the sketch function provided by the user
  // if a frame is drawn this function calls itself through window.requestAnimationFrame
  draw() {
    this.ctx.save();
    this.ctx.clearRect(0, 0, this.state.width, this.state.height);
    this.sketch(this);
    this.ctx.restore();

    if(this.ctx && !this.props.isPaused) {
      this.frameCount++;
      window.requestAnimationFrame(this.draw);
    }
  }

  // this function initiates the drawing loop
  startDraw() {
    this.frameCount = 0;
    window.requestAnimationFrame(this.draw);
  }


  componentDidUpdate(prevProps, prevState) {
    if(
      (prevProps.isPaused && !this.props.isPaused && this.ctx) ||
      (!prevState.ctx && this.state.ctx)
    ) {
      this.startDraw();
    }
  }

  componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();

    if(this.ctx && !this.props.isPaused) {
      this.startDraw();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  render() {
    return <canvas ref={this.onRefChange} width={this.state.width} height={this.state.height} style={{border: '1px solid #000'}}></canvas>;
  }
}
