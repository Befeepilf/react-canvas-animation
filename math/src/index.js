export const sin = Math.sin;
export const cos = Math.cos;
export const tan = Math.tan;
export const asin = Math.asin;
export const acos = Math.acos;
export const atan = Math.atan;
export const sqrt = Math.sqrt;
export const quad = b => Math.pow(b, 2);
export const pow = Math.pow;
