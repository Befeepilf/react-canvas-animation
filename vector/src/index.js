export {default} from './vector.js';
