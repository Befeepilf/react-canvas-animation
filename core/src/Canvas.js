import React from 'react';
import {init} from '@react-canvas-animation/draw';

export default function Canvas({isPaused, sketch}) {
  const [{width, height}, setCanvasSize] = React.useState({width: 0, height: 0});
  const [ctx, setCtx] = React.useState();
  const canvas = React.createRef();

  React.useEffect(() => {
    function onResize() {
      setCanvasSize({width: window.innerWidth, height: window.innerHeight});
    }

    onResize();

    window.addEventListener('resize', onResize);
    return () => window.removeEventListener('resize', onResize);
  }, []);

  React.useEffect(() => {
    let frameCount = 0;
    function internalDraw() {
      //console.log('draw', width)
      init(ctx, width, height);
      ctx.save();
      ctx.clearRect(0, 0, width, height);
      sketch(width, height, frameCount);
      ctx.restore();

      if(ctx && !isPaused) {
        frameCount++;
        window.requestAnimationFrame(internalDraw);
      }
    }

    if(ctx && !isPaused) {
      window.requestAnimationFrame(internalDraw);
    }
  }, [ctx, isPaused]);

  React.useEffect(() => {
    setCtx(canvas.current.getContext('2d'));
  }, [canvas.current]);


  return <canvas ref={canvas} width={width} height={height} style={{border: '1px solid #000'}}></canvas>;
}
