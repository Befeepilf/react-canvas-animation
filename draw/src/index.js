import Canvas from '@react-canvas-animation/core';

/*
  converts arguments to CSS compatible color values
  accepts either 1 or 3 arguments
  for 1 argument:
    if r is a string return r
    if r is a number treat g & b values as being equal to r
  for 3 arguments: threat them as red, green & blue values
*/
function argsToColor(r, g, b) {
  if(r !== undefined && g === undefined && b == undefined) {
    g = b = r;
  }
  return typeof r === 'string' ? r : `rgb(${r},${g},${b})`;
}

/*
  Adds a function to the prototype of Canvas
  To keep track of which custom functions has been added to its prototype add the function's name to an array in this prototype
  We need to keep track of added functions because inside Canvas they are lifted to the instance's this and
    Canvas is a React component so we only want to lift these custom functions and don't mess around with React specific functions
*/
function addToProto(name, func) {
  Canvas.prototype[name] = func;

  if(!Canvas.prototype.customFunctions) {
    Canvas.prototype.customFunctions = [];
  }
  Canvas.prototype.customFunctions.push(name);
}


addToProto('fill', function (r,g,b) {
  this.ctx.fillStyle = argsToColor(r,g,b);
});

addToProto('bg', function(r,g,b) {
  this.fill(r,g,b);
  this.ctx.fillRect(0, 0, this.width, this.height);
});

addToProto('stroke', function(r,g,b) {
  this.ctx.strokeStyle = argsToColor(r,g,b);
});

addToProto('strokeWeight', function(w) {
  this.ctx.lineWidth = w
});

addToProto('translate', function(x,y) {
  this.ctx.translate(x,y)
});

addToProto('rotate', function(rad) {
  this.ctx.rotate(rad);
});

addToProto('scale', function(x,y) {
  this.ctx.scale(x,y);
});

addToProto('line', function(x1,y1, x2,y2) {
  this.ctx.beginPath();
  this.ctx.moveTo(x1,y1);
  this.ctx.lineTo(x2,y2);
  this.ctx.stroke();
});

addToProto('rect', function(x,y,w,h) {
  this.ctx.fillRect(x,y,w,h);
});

addToProto('triangle', function(x1, y1, x2, y2, x3, y3) {
  this.ctx.beginPath();

  // if only 3 arguments are present
  // (x1, y1) = coordinate of the center of triangle's base
  // x2 = side length
  if(y2 === undefined && x3 === undefined && y3 === undefined) {
    this.ctx.moveTo(x1, y1);
    this.ctx.lineTo(x1 + x2 / 2, y1); // right corner
    this.ctx.lineTo(x1, y1 - x2 * Math.sqrt(3) / 2); // top corner
    this.ctx.lineTo(x1 - x2 / 2, y1); // left corner
  }
  // if all arguments are present
  else {
    this.ctx.moveTo(x1, y1);
    this.ctx.lineTo(x2, y2);
    this.ctx.lineTo(x3, y3);
  }
  this.ctx.fill();
});

addToProto('circle', function(x,y,r) {
  this.ctx.beginPath();
  this.ctx.arc(x, y, r, 0, 2 * Math.PI);
  this.ctx.fill();
});

addToProto('drawVector', function(v) {
  this.line(0, 0, v.x, v.y);
  this.ctx.save();
  this.translate(v.x, v.y);
  this.rotate(-v.angle() + Math.PI / 2);
  this.triangle(0, 0, 20);
  this.ctx.restore();
});
