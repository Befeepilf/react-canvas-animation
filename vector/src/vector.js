import {sin, cos, pow, sqrt} from '@react-canvas-animation/math';
import {line} from '@react-canvas-animation/draw';

export default class Vector {
  constructor(x,y) {
    this.x = x;
    this.y = y;
  }

  add(v) {
    this.x += v.x;
    this.y += v.y;
    return this;
  }

  sub(v) {
    this.x -= v.x;
    this.y -= v.y;
    return this;
  }

  mult(s) {
    this.x *= s;
    this.y *= s;
    return this;
  }

  div(s) {
    this.x /= s;
    this.y /= s;
    return this;
  }

  mag() {
    return sqrt(pow(this.x, 2) + pow(this.y, 2));
  }

  setMag(m) {
    this.div(this.mag()).mult(m);
    return this;
  }

  rotate(a) {
    const oldX = this.x;
    const oldY = this.y;
    this.x = oldX * cos(a) - oldY * sin(a);
    this.y = oldX * sin(a) + oldY * cos(a);
    return this;
  }

  draw() {
    line(0, 0, this.x, this.y);
  }

  copy() {
    return new Vector(this.x, this.y);
  }
}
