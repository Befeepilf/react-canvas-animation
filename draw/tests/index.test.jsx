import React from 'react';
import {render, waitFor} from '@testing-library/react';
import Canvas from '@react-canvas-animation/core';
import '../src';

describe("Set of drawing functions", () => {
    const drawingMethodNames = ['fill', 'bg', 'stroke', 'strokeWeight', 'translate', 'rotate', 'scale', 'line', 'rect', 'triangle', 'circle', 'drawVector'];

    it("is added to the canvas prototype", () => {
        const canvasMethodNames = Object.getOwnPropertyNames(Canvas.prototype);
        drawingMethodNames.forEach(methodName => {
            expect(canvasMethodNames.includes(methodName)).toBeTruthy();
        });
    });

    it("is exposed to the sketch function", async () => {
        let checked = false;
        render(<Canvas sketch={function(context) {
            drawingMethodNames.forEach(methodName => {
                expect(typeof context[methodName]).toEqual('function');
                expect(typeof this[methodName]).toEqual('function');
            });
            checked = true;
        }}/>);

        await waitFor(() => {
            expect(checked).toBeTruthy();
        });
    });


    describe("Individual drawing functions", () => {
        const {container, rerender} = render(<Canvas sketch={() => {}}/>);

        const getPixels = () => {
            const canvas = container.querySelector('canvas');
            const imgData = canvas.getContext('2d').getImageData(0, 0, canvas.width, canvas.height).data;
            const pixels = [];

            for(let i = 0; i < imgData.length; i += 4) {
                const pixel = {};

                pixel.r = imgData[i];
                pixel.g = imgData[i + 1];
                pixel.b = imgData[i + 2];
                pixel.a = imgData[i + 3];

                pixels.push(pixel);
            }

            return pixels;
        }

        it("fill", async () => {
            let fillStyle;
            let testStage = 0;

            await rerender(<Canvas sketch={({ctx, fill}) => {
                if(testStage === 0)
                    fill('red');
                else if(testStage === 1)
                    fill(56);
                else if(testStage === 2)
                    fill(27, 87, 102);

                fillStyle = ctx.fillStyle;
            }}/>);

            await waitFor(() => {
                expect(fillStyle).toEqual('#ff0000');
                testStage++;
            });

            await waitFor(() => {
                expect(fillStyle).toEqual('#383838');
                testStage++;
            });

            await waitFor(() => {
                expect(fillStyle).toEqual('#1b5766');
            });
        });

        it("bg", async () => {
            const width = 8;
            const height = 8;
            let hasDrawn = false;

            await rerender(<Canvas width={width} height={height} sketch={({bg}) => {
                bg('blue');
                hasDrawn = true;
            }}/>);

            await waitFor(() => {
                expect(hasDrawn).toBeTruthy();
            });

            getPixels().forEach(({r, g, b, a}) => {
                expect(r).toEqual(0);
                expect(g).toEqual(0);
                expect(b).toEqual(255);
                expect(a).toEqual(255);
            });
        });
    });
});