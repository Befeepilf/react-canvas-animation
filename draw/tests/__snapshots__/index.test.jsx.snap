// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set of drawing functions Individual drawing functions circle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAABmJLR0QA/wD/AP+gvaeTAAAADklEQVQYlWNgGAWDEwAAAZoAARbK02kAAAAASUVORK5CYII="`;

exports[`Set of drawing functions Individual drawing functions drawVector 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAADiUlEQVR4nO3by2tcVRwH8E+TKi2tRWyRiIgvFLEoIhZFxEJFEFSKdFHcdSG+oLr0X1Bw48KtCxcudFEURAWRokQRFEoVUXyASOuzikh9a1ychAlx0kwmM/ccZr4fuEyYBPKFH2fuvXPPF+ZFM2Ywh5tqB4lidvH1IF6oGSR6tuE7XFY7SJQV8hd24ma8VjdOLLkQP+Dc2kGm3dI55Bdcowzm7XpxYrlr8RXOqh1kms0u+/lb3I3f8WGdOLHSXXivdojo2YSPsLd2kOh5AC/WDhE9W/A1rqodZBrN9nnvb+zAPrzcbZxYzfk4hV21g0ybfisETuNKXI63uosTZ7Ib3yjnlOjIaisEvsdtyqXwsW7ixFpuxwfKUKIRx5TBRCMO4ZXaIaLnbJxUvg2OMTvTSX3JP8qV1h14abxxYlDnKTeKc7WDTLpBVgj8hktwNY6OK0yszxXKvcm22kEm2aArBH7ELdiK98cTJ9ZrLz5RdjxGI95VHvVGI+7FG7VDRM9mfIk9tYNMovWc1Jf8qwxlP46MNk4Ma4ey9fSi2kEmzTArBP7ABbger48uTmzExcqN4vbaQSbJsCsEflZO7DuVS+FowB58YWODjRGbx4HaIaLngDR5mzKLz6TJOxKj+OxfWHxNk7chafKOyKiujtLkbVCavCMwyvuHNHkblCbvBo36DjtN3galyduYNHkblCZvY9LkHdK4vjZPk7dBafIOYZwPltLkbVCavOs07kevafI2KE3eBqXJ25hD0uRtSpq8A+pqP1WavA1Kk3cAXe44TJO3QWnyrqHrPblp8jYoTd4GpcnbmDR5G5Mm7ypqFW3S5G1Qmrx91KyiDdPkndGrP8QYDNLknVF2Q74jG7k78Twe6fP+OXhUOfkv4NcuQ02zlU3eS/E4flIGsXR8WiXdlJrHw3hWKf8s9DmO1grXpc2V//8M7lx8fXqNvz05/jjTazvux8f6r4Z+x5NVknas6xUyhwdxWHlgtR5ZISN0nXJ++NPgK2LlcbDz1BNqBk8oz9WHHcYCbu06+KTbhxOGH0g68GOwS9l1MsxAtlbIOxU2KVdYpw0+jFNVkk6Z3ThusIEcr5Rx6mzBU9YeyKu1Ak6re5SPpdUG8ky9aN1qZefHEeVe5c1Vfn+iwyyxzCwe8/8byIdqhgpuxOd6A9lfN05Qnrk/pwzkhspZYpn7ZMd8RERERETE2PwHEeD+5dQvrGYAAAAASUVORK5CYII="`;

exports[`Set of drawing functions Individual drawing functions triangle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAsAAAALCAYAAACprHcmAAAABmJLR0QA/wD/AP+gvaeTAAAAcUlEQVQYlaXKsQnCQABA0Q9mgOAAQjqDK6RwG50mGcTUKZzCwikCt4CBfBsPQvC4QP5vH+xO6q3whASk2oJ7ROSRgw0y/7DINQUPyGsBRd5I8Q/fVzB+W8MSGRM4IMcl7hIw3kZ4Rj4ZPCEXkCED488vfcd2Bn04S9YAAAAASUVORK5CYII="`;
